projects/baby_driver/test/test_i2c_write.c

```diff
@@ -0,0 +1,247 @@
+  uint8_t command_1[8] = { BABYDRIVER_MESSAGE_I2C_WRITE_COMMAND, I2C_PORT_1, 0x74, 14, 2, 3, 0, 0 };
+  uint8_t data_1[8] = { BABYDRIVER_MESSAGE_I2C_WRITE_DATA, 10, 22, 5, 4, 3, 76, 120 };
+  uint8_t data_2[8] = { BABYDRIVER_MESSAGE_I2C_WRITE_DATA, 43, 2, 3, 4, 5, 1, 8 };
+
+  CAN_TRANSMIT_BABYDRIVER(command_1[0], command_1[1], command_1[2], command_1[3], command_1[4],
+                          command_1[5], command_1[6], command_1[7]);
+
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+  MS_TEST_HELPER_ASSERT_NO_EVENT_RAISED();
+
+  CAN_TRANSMIT_BABYDRIVER(data_1[0], data_1[1], data_1[2], data_1[3], data_1[4], data_1[5],
+                          data_1[6], data_1[7]);
+
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+  MS_TEST_HELPER_ASSERT_NO_EVENT_RAISED();
+
+  CAN_TRANSMIT_BABYDRIVER(data_2[0], data_2[1], data_2[2], data_2[3], data_2[4], data_2[5],
+                          data_2[6], data_2[7]);
+
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+
+  TEST_ASSERT_EQUAL(1, s_times_callback_called);
+  TEST_ASSERT_EQUAL(BABYDRIVER_MESSAGE_STATUS, s_received_data[0]);
+  TEST_ASSERT_EQUAL(s_status_return, s_received_data[1]);
+
+  TEST_ASSERT_EQUAL(I2C_PORT_1, i2c_test);
+  TEST_ASSERT_EQUAL(0x74, addr_test);
+  TEST_ASSERT_EQUAL(3, reg_test);
+  TEST_ASSERT_EQUAL(14, tx_len_test);
+  for (uint8_t i = 0; i < 7; i++) {
+    TEST_ASSERT_EQUAL(data_1[i + 1], tx_data_test[i]);
+  }
+  for (uint8_t j = 0; j < 7; j++) {
+    TEST_ASSERT_EQUAL(data_2[j + 1], tx_data_test[j + 7]);
+  }
+
+  // Test receiving multiple messages and writing over i2c without register
+  uint8_t command_2[8] = { BABYDRIVER_MESSAGE_I2C_WRITE_COMMAND, I2C_PORT_2, 0x55, 14, 0, 0, 0, 0 };
+
+  CAN_TRANSMIT_BABYDRIVER(command_2[0], command_2[1], command_2[2], command_2[3], command_2[4],
+                          command_2[5], command_2[6], command_2[7]);
+
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+  MS_TEST_HELPER_ASSERT_NO_EVENT_RAISED();
+
+  CAN_TRANSMIT_BABYDRIVER(data_1[0], data_1[1], data_1[2], data_1[3], data_1[4], data_1[5],
+                          data_1[6], data_1[7]);
+
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+  MS_TEST_HELPER_ASSERT_NO_EVENT_RAISED();
+
+  CAN_TRANSMIT_BABYDRIVER(data_2[0], data_2[1], data_2[2], data_2[3], data_2[4], data_2[5],
+                          data_2[6], data_2[7]);
+
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+  MS_TEST_HELPER_CAN_TX_RX(TEST_CAN_EVENT_TX, TEST_CAN_EVENT_RX);
+
+  TEST_ASSERT_EQUAL(2, s_times_callback_called);
+  TEST_ASSERT_EQUAL(BABYDRIVER_MESSAGE_STATUS, s_received_data[0]);
+  TEST_ASSERT_EQUAL(s_status_return, s_received_data[1]);
+
+  TEST_ASSERT_EQUAL(I2C_PORT_2, i2c_test);
+  TEST_ASSERT_EQUAL(0x55, addr_test);
+  TEST_ASSERT_EQUAL(0, reg_test);
```

Comment:
This test will actually pass even if your code calls `i2c_write_reg` when `is_reg` is 0, since you set `reg_test = 0` in your `i2c_write` mock and you passed 0 in the `reg` field in the command_2 message here. You need another way to differentiate - maybe pass a nonzero `reg` in command_2.